src/librbd/io/ImageRequest.cc

```diff
@@ -234,20 +234,24 @@ void ImageReadRequest<I>::send_request() {
   }
   
   utime_t ts = ceph_clock_now();
-  bool update_timestamp =
-    (image_ctx.atime_update_interval && (static_cast<uint64_t>(ts.sec())
-    >= image_ctx.atime_update_interval + image_ctx.get_access_timestamp()))
-    ? true : false;
+  bool update_timestamp = false;
+  {
+    RWLock::WLocker access_timestamp_locker(image_ctx.access_timestamp_lock);
```

Comment:
Nit: take the lock for read first, check to see if you pass the throttle. If no, drop the lock and continue. Otherwise, drop the lock and take a writer lock and re-check:

```
namespace {

bool should_update_timestamp(const utime& now, const utime_t& last_update, uint64_t interval) {
  return ((interval > 0) && (now >= last_update + interval));
}

} // anonymous namespace

... snip ...

ImageReadRequest<I>::send_request() {
... snip ...
image_ctx.timestamp_lock.get_read();
if (should_update_timestamp(ts, image.get_access_timestamp(),
                            image_ctx.atime_update_interval)) {
  image_ctx.timestamp_lock.put_read();
  image_ctx.timestamp_lock.get_write();
  if (should_update_timestamp(ts, image.get_access_timestamp(),
                            image_ctx.atime_update_interval)) {
    image_ctx.set_access_timestamp(ts);
  }
  image_ctx.timestamp_lock.put_write();
} else {
  image_ctx.timestamp_lock.put_read();
}

```

Reply:
Thanks for the feedback! I tried something in commit #27b7529 - would be great to have your thoughts on it.

Reply:
That looks good to me -- just need it on the write path as well.